src/value.rs

```diff
@@ -0,0 +1,553 @@
+
+                Err(e) => return Err(ParseError::from(e))?,
+
+                _ => return Err(ParseError::UnexpectedEvent("member".into()))?,
+            }
+        }
+
+        Ok(ret.into())
+    }
+
+    fn read_base64_from_reader<K: AsRef<[u8]>>(
+        end: K,
+        reader: &mut Reader<&[u8]>,
+        buf: &mut Vec<u8>,
+    ) -> Result<Self> {
+        let text = reader.read_text(end, buf).map_err(ParseError::from)?;
+        Ok(decode_base64(&text).map_err(ParseError::from)?.into())
+    }
+
+    fn read_array_from_reader<K: AsRef<[u8]>>(
+        end: K,
+        mut reader: &mut Reader<&[u8]>,
+        mut buf: &mut Vec<u8>,
+    ) -> Result<Self> {
+        let mut ret: Vec<Self> = Vec::new();
+
+        // The inner tag of an array should be a data tag.
+        reader.expect_tag(b"data", buf)?;
+
+        // Read the next event. If it's text or the value closing tag, we know
+        // we've got a string. If it's a start tag, we've got more work to do.
+        loop {
+            match reader.read_event(&mut buf) {
+                // If we got the matching end data tag, we need to exit the loop
+                // so we can handle cleanup.
+                Ok(Event::End(ref e)) if e.name() == b"data" => {
+                    break;
+                }
+
+                // If we got a start tag, we need to handle each of the value types.
+                Ok(Event::Start(ref e)) => match e.name() {
+                    b"value" => ret.push(Value::read_value_from_reader(&mut reader, &mut buf)?),
+                    _ => {
+                        return Err(ParseError::UnexpectedTag(
+                            String::from_utf8_lossy(e.name()).into(),
+                            "value".into(),
+                        ))?;
+                    }
+                },
+
+                // Possible error states
+                Ok(Event::Eof) => return Err(ParseError::UnexpectedEOF("value".into()))?,
+
+                Err(e) => return Err(ParseError::from(e))?,
+
+                _ => return Err(ParseError::UnexpectedEvent("value".into()))?,
+            }
+        }
+
+        reader
+            .read_to_end(end, &mut buf)
+            .map_err(ParseError::from)?;
+
+        Ok(ret.into())
+    }
+
+    fn read_nil_from_reader<K: AsRef<[u8]>>(
+        end: K,
+        reader: &mut Reader<&[u8]>,
+        buf: &mut Vec<u8>,
+    ) -> Result<Self> {
+        reader.read_to_end(end, buf).map_err(ParseError::from)?;
+        Ok(Self::Nil)
+    }
+}
+
+impl From<i32> for Value {
+    fn from(other: i32) -> Self {
+        Value::Int(other)
+    }
+}
+
+impl From<i64> for Value {
+    fn from(other: i64) -> Self {
+        Value::Int64(other)
+    }
+}
+
+impl From<bool> for Value {
+    fn from(other: bool) -> Self {
+        Value::Bool(other)
+    }
+}
+
+impl From<String> for Value {
+    fn from(other: String) -> Self {
+        Value::String(other)
+    }
+}
+
+impl<'a> From<&'a str> for Value {
```

Comment:
Not sure you need lifetimes here.  Did the compiler squawk at you?

The compiler should be able to infer the lifetime of 'other' in this situation.  And since to_string() should be cloning the contents of the slice, you shouldn't need to tie the lifetime of 'Self' to the lifetime of 'other'.

Reply:
Yeah, I think this is leftover from when I was playing around with storing the &str rather than a String to avoid allocations.

Reply:
Yeah, I think this is leftover from when I was playing around with storing the &str rather than a String to avoid allocations.